typing/typetexp.ml

```diff
@@ -50,8 +51,259 @@ type error =
+  val new_any_var : Location.t -> Env.t -> policy -> type_expr
+    (* create a new variable to represent a _; fails for fixed_policy *)
+  val new_var : ?name:string -> policy -> type_expr
+    (* create a new variable according to the given policy *)
+
+  val add_pre_univar : type_expr -> policy -> unit
+    (* remember that a variable might become a univar if it isn't unified;
+       used for checking method types *)
+
+  val collect_univars : (unit -> 'a) -> 'a * type_expr list
+    (* collect univars during a computation; returns the univars.
+       The wrapped computation should use [univars_policy].
+       postcondition: the returned type_exprs are all Tunivar *)
+
+  val reset_locals : ?univars:poly_univars -> unit -> unit
+    (* clear out the local type variable env't; call this when starting
+       a new e.g. type signature. Optionally pass some univars that
+       are in scope. *)
+
+  val lookup_local : string -> type_expr
+    (* look up a local type variable; throws Not_found if it isn't in scope *)
+
+  val remember_used : string -> type_expr -> Location.t -> unit
+    (* remember that a given name is bound to a given type *)
+
+  val globalize_used_variables : globals_only:bool -> Env.t ->
+    fixed:bool -> unit -> unit
+    (* after finishing with a type signature, add used variables to the
+       global type variable scope; with globals_only, only already-in-scope
+       variables are considered (but they are still unified with the global
+       type variables *)
+
+end = struct
+  (** Map indexed by type variable names. *)
+  module TyVarMap = Misc.Stdlib.String.Map
+
+  let not_generic v = get_level v <> Btype.generic_level
+
+  (* These are the "global" type variables: they were in scope before
+     we started processing the current type.
+
+     INVARIANT: No types at generic_level.
+  *)
+  let type_variables = ref (TyVarMap.empty : type_expr TyVarMap.t)
+
+  (* These are variables that have been used in the currently-being-checked
+     type.
+
+     INVARIANT: No types at generic_level.
+  *)
+  let used_variables =
+    ref (TyVarMap.empty : (type_expr * Location.t) TyVarMap.t)
+
+  (* These are variables we expect to become univars (they were introduced with
+     e.g. ['a .]), but we need to make sure they don't unify first.  Why not
+     just birth them as univars? Because they might successfully unify with a
+     row variable in the ['a. < m : ty; .. > as 'a] idiom.  They are like the
+     [used_variables], but will not be globalized in [globalize_used_variables].
+
+     INVARIANT: No types at generic_level.
+  *)
+  let univars = ref ([] : (string * type_expr) list)
+  let assert_univars uvs =
+    assert (List.for_all (fun (_name, v) -> not_generic v) uvs)
+
+  (* These are variables that will become univars when we're done with the
+     current type. Used to force free variables in method types to become
+     univars.
+
+     INVARIANT: No types at generic_level.
```

Comment:
I don't think so. It looks to me that the only place this list is extended is in `add_pre_univar` which asserts that the type isn't generic.

Reply:
Same as above, the call to `generalize` in `transl_simple_type_univars` may lift type variables to the generic level, and those lifted type variables are precisely the one collected by `collect_univars`. 

I was making the remark that this invariant is quite close to be true since it seems to be only temporarily broken between this call to `generalize` and the end of the collection of `pre_univars` in `collect_univars` and I was wondering if we could adjust the condition to be a full invariant. 

Sorry for being unclear!